hbase-server/src/main/java/org/apache/hadoop/hbase/master/cleaner/CleanerChore.java

```diff
@@ -423,16 +426,18 @@ protected Boolean compute() {
         allFilesDeleted = deleteAction(() -> checkAndDeleteFiles(files), "files");
       }
 
+      List<CompletableFuture<Boolean>> futures = new ArrayList<>();
       boolean allSubdirsDeleted = true;
       if (!subDirs.isEmpty()) {
-        List<CleanerTask> tasks = Lists.newArrayListWithCapacity(subDirs.size());
         sortByConsumedSpace(subDirs);
-        for (FileStatus subdir : subDirs) {
-          CleanerTask task = new CleanerTask(subdir, false);
-          tasks.add(task);
-          task.fork();
-        }
-        allSubdirsDeleted = deleteAction(() -> getCleanResult(tasks), "subdirs");
+        subDirs.forEach(dir -> futures.add(pool.execute(new CleanerTask(dir, false))));
```

Comment:
I think submitting the task to the same pool is not safe here. In ForkJoinPool, it is safe to call join as in the join method, the current thread will try to execute tasks if there are any, so in general, we can always make progress and finally finish all the tasks. But here, the current thread will be blocked on the future.get, so it is possible that all the threads in the pool are blocked on a future.get and cause dead lock here.